lib/scheduler.ml

```diff
@@ -428,23 +430,44 @@ module Make (Streamd : StreamDescriptor) = struct
           (* We check for activity again, because the stream may have gone
            * inactive after the call to `write` above. *)
           written, Streamd.requires_output descriptor
-        else (
-          match PriorityQueue.pop p.children with
-          | Some ((id, (Stream i as i_node)), children') ->
-            p.t_last <- i.t;
-            let written, subtree_is_active = schedule i_node in
-            if subtree_is_active
-            then (
-              update_t i_node written;
-              p.children <- PriorityQueue.add id i_node children')
-            else (
-              implicitly_close_idle_stream i.descriptor max_seen_ids;
-              p.children <- children');
-            written, subtree_is_active
-          | None ->
-            (* Queue is empty, see line 6 above. *)
-            0, false)
+        else traverse p_node children
+    and traverse : type a. a node -> PriorityQueue.t -> int * bool =
+     fun p_node ->
+      let rec loop children =
+        match PriorityQueue.pop children with
+        | Some ((id, (Stream i as i_node)), children') ->
+          let written, subtree_is_active = schedule i_node in
+          if not subtree_is_active
+          then (
+            implicitly_close_idle_stream i.descriptor max_seen_ids;
+            (* XXX(anmonteiro): we may not want to remove from the tree right
+             * away. *)
+            remove_child p_node id);
+          if written > 0
+             (* We need this check because the queue contains both streams that
+              * want to send data and streams that don't want to send data.
+              * Without this check, we might end up in a situation where a stream
+              * won't be flushed until a few write operations later. *)
+          then (
+            update_t_last p_node i.t;
+            update_t i_node written;
```

Comment:
in the previous code we were adding `i_node` back to the priority queue of `children'`. This is important if `t` has changed in order to re-balance the tree (just mutating `t` won't do it.)

Reply:
you should be able to write a test that fails with the current code because of this.

Reply:
even trickier: we should only put the stream back in the tree if the subtree isn't active.

Reply:
> in the previous code we were adding `i_node` back to the priority queue of `children'`. This is important if `t` has changed in order to re-balance the tree (just mutating `t` won't do it.)

Of course you're right 🤦 
I introduced that regression in my latest commit when I though I could "simplify" things. I reverted this commit although I understand you're fixing it in #204.